File: typescript/binary-search/binary-search.ts
export function find(haystack: number[], needle: number): number | never {
  let left: number = 0;
  let right: number = haystack.length - 1;
  while (left <= right) {
    let mid = Math.floor((right + left) / 2);
    let diff = haystack[mid] - needle;
    if (diff == 0) {
      return mid;
    } else if (diff < 0) {
      left = mid + 1
    } else {
      right = mid - 1
    }
  }
  throw new Error('Value not in array');
}

File: c/binary-search/binary_search.c
#include "binary_search.h"
#include <stdio.h>

const int *binary_search(int value, const int *arr, size_t length) {
   int left = 0;
   int right = length - 1;
   while (left <= right) {
      int mid = (left + right) / 2;
      if (arr[mid] == value) {
         return &arr[mid];
      }
      if (arr[mid] < value) {
         left = mid + 1;
      } else {
         right = mid - 1;
      }
   }
   return NULL;
}
